pennylane/measurements.py

```diff
@@ -125,6 +152,121 @@ def __init__(self, return_type, obs=None, wires=None, eigvals=None, id=None):
         # Queue the measurement process
         self.queue()
 
+    def shape(self, device=None):
+        """The expected output shape of the MeasurementProcess.
+
+        Note that the output shape is dependent on the device when:
+
+        * The return_type is either Probability, State (from qml.state) or
+          Sample;
+        * The shot vector was defined in the device.
+
+        Args:
+            device (.Device): a PennyLane device to use for determining the
+                shape
+
+        Returns:
+            tuple: the output shape
+        """
+        if device is None:
+            if self._shape is not None:
+                return self._shape
+
+            if self.return_type in (Probability, State, Sample):
+                raise MeasurementShapeError(
+                    f"Return type {self.return_type} requires the device argument to be passed to obtain the shape."
+                )
+
+        if device._shot_vector is None:
+
+            # Device dependent shapes
+            if self.return_type == Probability:
+                len_wires = len(self.wires)
+                dim = self._get_num_basis_states(len_wires, device)
+                shape = (1, dim)
+
+            elif self.return_type == State and self._shape is None:
+
+                # Note: qml.density_matrix has its shape defined, so we're handling
+                # the qml.state case; acts on all device wires
+                dim = 2 ** len(device.wires)
+                shape = (1, dim)
+
+            elif self.return_type == Sample:
+                len_wires = len(device.wires)
+
+                if self.obs is not None:
+                    # qml.sample(some_observable) case
+                    shape = (1, device.shots)
+
+                else:
+                    # qml.sample() case
+                    shape = (1, device.shots, len_wires)
+            else:
+                shape = self._shape
+
+            # The other return types should have their shapes pre-defined
+            return shape
+
+        return self._shot_vector_shape(device)
+
+    def _shot_vector_shape(self, device):
+        """Auxiliary function for getting the output shape when the device has
+        the shot vector defined.
+
+        The shape is device dependent even if the return type has a shape
+        pre-defined (e.g., expectation values, states, etc.).
+        """
+        shot_vector = device._shot_vector
+        num_shot_elements = sum([s.copies for s in shot_vector])
+
+        if self._shape is not None:
+
+            shape = (num_shot_elements,)
```

Comment:
Should it not contain the `self._shape` in the second part of the tuple?

Reply:
Great catch! This def is good for expval and var, but doesn't work for `qml.state` or `qml.density_matrix`. For the latter, made updates in https://github.com/PennyLaneAI/pennylane/pull/2044/commits/c1a32b5bc63071fcea3f8068df1ca0e080af5bf0.